WIP: Virtual sensors / remap sensors

mainakjas opened the pull request:
Closes https://github.com/mne-tools/mne-python/issues/1080. Here's what you get when you remap mag to grad for sample.

![virtual_channels](https://cloud.githubusercontent.com/assets/3817535/7093819/c7c7d004-dfbc-11e4-978a-1ed71c0f6289.png)
Still needs to be made nicer with tests and better options in API.

cc @Anevar @agramfort

mainakjas commented:
We should think of a good way to convert the info structure -- otherwise you'll get plots with units all wrong. If we convert 204 grads to 204 mags, how do we specify the channel names of the new virtual channels?

agramfort commented on examples/plot_virtual_evoked.py:
```diff
@@ -0,0 +1,33 @@
+evoked = read_evokeds(evoked_fname, condition=condition, baseline=(-0.2, 0.0))
+
+virtual_evoked = compute_virtual_evoked(evoked, from_type=from_type,
+                                        to_type=to_type)
```
you should keep only the grads in evoked and convert it to mags and plot the true mags and the generated ones. That would be a great unit test too

Eric89GXL replied:
+1 this would be great, I'm really curious how closely it will match

mainakjas replied:
going from grad to mags appears to work fine

![grad_to_mag](https://cloud.githubusercontent.com/assets/3817535/7103798/819791e8-e0b9-11e4-819b-89581d392a88.png)

but the other way round (mag to grad) looks a bit suspicious to me

![mag_to_grad](https://cloud.githubusercontent.com/assets/3817535/7103801/8c8d6f64-e0b9-11e4-9992-67b5191a7ca0.png)

Fig 1 is True mag/grad whereas Fig 2 is interpolated

Eric89GXL replied:
I wonder if it has to do with the nature of the neural generators and sensitivity of the different sensors to those sources. It could also be that the mags are picking up more external interference (I think they should), and the grads wouldn't be as sensitive to that.

If you have lots of free time you could simulate different sources and see which ones interpolate well grad->mag, mag->grad, or both, but it's probably overkill.

@agramfort WDYT?

mainakjas replied:
correction: Fig 1 is interpolated whereas Fig 2 is True mag/grad

mainakjas replied:
Btw, the magnitude also looks off when we go from mag to grad. Could it have something to do with the fact that grads measure difference in flux?

Eric89GXL replied:
Could be. We might need to dig into `mne_map_data` to get it right. Actually, you might be able to use `mne_map_data` as a benchmark.

mainakjas replied:
ok, we have the signs thanks to brainstorming with @teonlamont and @ybekhti. The grad topomap had the RMS values of two pairs, so for interpolated 'grad'. we take the absolute magnitude (since we don't have any pairs) for better comparison:

![mag_to_grad](https://cloud.githubusercontent.com/assets/3817535/7118017/ce82a328-e1fc-11e4-91c4-fda2b0bc5b90.png)

Fig 1 is interpolated grad and Fig 2 is true grad. 

There is still a weird magnitude bias which needs to be figured out. When going from grad to mag the interpolated mag data is higher in magnitude compared to True mag whereas when going from mag to grad the interpolated grad data is lower in magnitude compared to True grad

Eric89GXL replied:
Is it off by a factor of about 100? I've seen that number floating around before.

mainakjas replied:
It's more like a factor of 20 ...

Eric89GXL replied:
@mainakjas do you have time to dig into `mne_map_data`? Maybe tackle the other PRs related to `report` before you do :)

mainakjas replied:
@Eric89GXL the `report` PR will progress when I meet @dengemann . In the meanwhile, I wonder if this is to blame : https://github.com/mainakjas/mne-python/blob/master/mne/forward/_lead_dots.py#L144 for problems with mag -> grad conversion. How do I modify it for grads?

teonlamont commented on mne/forward/_field_interpolation.py:
```diff
@@ -87,6 +89,87 @@ def _compute_mapping_matrix(fmd, info):
+    ----------
+    evoked : instance of mne.Evoked
+        The evoked object.
+    from_type : str
```
why is `from` the destination and `to` origin?

mainakjas replied:
my bad, fixed now :)

dengemann commented:
@agramfort @mainakjas take a look here:

https://github.com/fieldtrip/fieldtrip/blob/master/template/layout/neuromag306cmb.lay

combined grads + mags layout

agramfort commented:
how is that related? the lay does not tell you the 3d locations of sensors
+ their orientation

dengemann commented:
> the lay does not tell you the 3d locations of sensors

that's indeed true.

mainakjas commented:
These virtual channels create a discrepancy between the channel names and channel type which causes a lot of code to break. How do we tell the `evoked` methods that these are virtual channels and should be treated differently?

mainakjas commented:
I can probably make it work if we make new channel types `vmag` and `vgrad` -- then we treat these differently in all the `pick` methods. Thoughts?

agramfort commented:
-1 for new types. The units match the meg ch type

Eric89GXL commented:
I agree on the new types, the idea is that we are interpolating/synthesizing/simulating data that could/would have been recorded on real, standard sensors. We could give the new channels distinctive names like `MEGInterp001` by default so they know the data are interpolated/synthesized, and strongly emphasize on the docstring that data are synthesized/interpolated.

Eric89GXL commented:
I think a simple naming scheme by default `'%s_interp_%s' % (ch_type.upper(), idx)` would be a good scheme if users pass `new_ch_names=None`. If they want something more compact, they should override the default verbose name.

Eric89GXL commented:
@mainakjas this seems to be close to working, I added the 0.9 tag in hopes that you can get this done in the coming week. Sound feasible?

mainakjas commented:
@Eric89GXL I think we can have it ready in the coming week definitely

agramfort commented:
You need to change the coil type

mainakjas commented:
I made it work for grads too. Don't ask my how or why. This is for somato data. Have a look:

![virt_grads_working](https://cloud.githubusercontent.com/assets/3817535/7140572/4f4ff3c8-e2cc-11e4-81ae-06462cd7753c.png)

mainakjas commented:
I made it work for grads too. Don't ask my how or why. This is for somato data. Have a look:

![virt_grad_working](https://cloud.githubusercontent.com/assets/3817535/7140766/838ef5f2-e2cd-11e4-987b-da32b8ded7db.png)

Eric89GXL commented:
Wow, those are actually pretty close. I see your hack in the code, I too am confused by it :(

mainakjas commented:
ok hack doesn't work any more @agramfort It seems I had misunderstood the eventual goal. When the 'to_type' was 'grad', the interpolation was being done on the original 'mag' channel locations using the 'grad' data. However, what we want is to interpolate at the 'grad' channel locations using the mag data. This turns out to be more challenging somehow :) It still works for grad to mag but not the other way round.

![better_understanding](https://cloud.githubusercontent.com/assets/3817535/7165637/13045552-e3a9-11e4-9cf2-25b284fc9615.png)

The interpolated grad maps basically look like the mags. Changing the coil type is not going to do anything -- I've tried all possible combinations. I think now I need to understand what's going on :) @Eric89GXL any hints?

Eric89GXL commented:
I'd take a look at `mne_map_data` to figure out what it does. It's possible that the field interpolation code probably only into magnitude sensors, not gradient sensors. `mne_map_data` should have to do both, though.

mainakjas commented:
so `mne_map_data` is not part of mne-python code? It's in the C code, right? That's why I was confused when you first mentioned it. I'll dig into it tomorrow. Thanks!

Eric89GXL commented:
Yep, you'll need to dig through the C code.

In some (hopefully many) of the functions I translated, I tried to note function names or filenames that they were drawn from. My convention has been to drop the `mne` from in front of things, turning e.g. `mne_compute_forward` into `_compute_forward`. That might help you find correspondences between code you look at and code in mne-python. Feel free to beef up the mne-python comments as you go if it helps you navigate the C code -- I tried to put enough in there to be able to follow, so you looking through the two is a nice test case.

Eric89GXL commented:
Nice additional comments @mainakjas. Let me know if you need help, I might have time next week to dig into this, too. I have a feeling this PR might end up holding up the release, the others seem more straightforward. But it's probably worth it to get this in. WDYT @agramfort @dengemann?

mainakjas commented:
I think I'm on to something (not sure). I'll try for one or two more days. If it does not work, @agramfort suggested that we could simply do it for `mags` for now and extend it to `grads` later.

Eric89GXL commented:
Yes but that would be so unsatisfying :)

mainakjas commented:
Holy cow, I just translated the C code and it works like magic. Pushing the changes now.

agramfort commented:
the C code is readable after all :)

mainakjas commented:
![both_work](https://cloud.githubusercontent.com/assets/3817535/7208336/55712936-e540-11e4-8aeb-ecf012ea54ac.png)

agramfort commented:
tell us when to review ie you have tests + travis happy

mainakjas commented:
yeah, I gave up on installing all those libraries. I just read it and translated it :)

teonlamont commented on mne/forward/_field_interpolation.py:
```diff
@@ -87,6 +91,127 @@ def _compute_mapping_matrix(fmd, info):
+    # pick the original and destination channels
+    pick_from = pick_types(evoked.info, meg=from_type, eeg=False,
+                           ref_meg=False)
+    pick_to = pick_types(evoked.info, meg=to_type, eeg=False,
```
to use these picks, don't you need to check that the `pos` from `picks_from` is equal to `pos` from `picks_to`?

mainakjas replied:
by `pos` you mean channel positions? why would you need the `from` and `to` channel positions to be the same? At least, what I am doing right now is this: If we go from say, 'grad' to 'mag', we take the `grad` data and transform it into 104 channels of `mag` data. Ultimately, we want to go from 306 channels (grad + mag) into either 204 channels of 'grad' or 102 channels of 'mag'.

Eric89GXL replied:
In my mind the goal is to go from `N` sensors of arbitrary types at arbitrary locations, and interpolate to `M` sensors of arbitrary types at arbitrary locations. I think your code can be structured to allow this, in which case what you describe are just particular use cases...

agramfort replied:
a big use case is also to "repair" bad MEG channels with proper
interpolation.

Eric89GXL replied:
Exactly -- another particular use case of the general idea I mentioned above. I think that should also be part of the release, even if it means delaying a week or two to get it in. @mainakjas do you have the bandwidth to really work on this in the coming week so we can get the generalization working, and then the particular use cases? I can see three: virtual channels (done), proper interpolation, and mapping data between head positions. The last two should be fairly straightforward if the generalization is good.

My vote is to try to get all three done, and if it ends up being too ambitious, we can scale back later.

mainakjas replied:
Sounds good. Do you want to do this in three separate PRs or a single PR? I'd prefer separate PRs -- we can refactor along the way. That way, even if we are done 2 out of 3 applications, we don't get stuck because of a mega PR. Wdyt? I beefed up the comments significantly based on my understanding. I still don't understand all the physics but maybe @agramfort can help with my questions when he's back in Paris :)

Eric89GXL replied:
I can live with separate ones, let me know when this one is good to go, then

Eric89GXL replied:
(we an assume Matti got the physics right for now :) )

mainakjas replied:
On Sat, Apr 18, 2015 at 1:54 AM, Eric Larson <notifications@github.com>
wrote:

> In mne/forward/_field_interpolation.py
> <https://github.com/mne-tools/mne-python/pull/1943#discussion_r28638165>:
>
> > +    if from_type not in ['mag', 'grad']:
> > +        raise ValueError('from_type must be "mag" or "grad", not "%s"'
> > +                         % from_type)
> > +    if to_type not in ['mag', 'grad']:
> > +        raise ValueError('to_type must be "mag" or "grad", not "%s"'
> > +                         % to_type)
> > +    if from_type == to_type:
> > +        raise ValueError('from_type and to_type cannot be the same.')
> > +
> > +    # get surface in head coordinates
> > +    surf_virt = get_meg_helmet_surf(evoked.info)
> > +
> > +    # pick the original and destination channels
> > +    pick_from = pick_types(evoked.info, meg=from_type, eeg=False,
> > +                           ref_meg=False)
> > +    pick_to = pick_types(evoked.info, meg=to_type, eeg=False,
>
> (we an assume Matti got the physics right for now :) )
>
of course :)

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1943/files#r28638165>.
>

teonlamont replied:
you're constructing the info using the `pick_to` info so I was trying to figure out why. the virtual_epochs has the shape of your `pick_from`.

mainakjas replied:
no, virtual evoked has shape of `pick_to` :)

mainakjas replied:
On Fri, Apr 17, 2015 at 10:07 PM, Eric Larson <notifications@github.com>
wrote:

> In mne/forward/_field_interpolation.py
> <https://github.com/mne-tools/mne-python/pull/1943#discussion_r28625950>:
>
> > +    if from_type not in ['mag', 'grad']:
> > +        raise ValueError('from_type must be "mag" or "grad", not "%s"'
> > +                         % from_type)
> > +    if to_type not in ['mag', 'grad']:
> > +        raise ValueError('to_type must be "mag" or "grad", not "%s"'
> > +                         % to_type)
> > +    if from_type == to_type:
> > +        raise ValueError('from_type and to_type cannot be the same.')
> > +
> > +    # get surface in head coordinates
> > +    surf_virt = get_meg_helmet_surf(evoked.info)
> > +
> > +    # pick the original and destination channels
> > +    pick_from = pick_types(evoked.info, meg=from_type, eeg=False,
> > +                           ref_meg=False)
> > +    pick_to = pick_types(evoked.info, meg=to_type, eeg=False,
>
> Exactly -- another particular use case of the general idea I mentioned
> above. I think that should also be part of the release, even if it means
> delaying a week or two to get it in. @mainakjas
> <https://github.com/mainakjas> do you have the bandwidth to really work
> on this in the coming week so we can get the generalization working, and
> then the particular use cases? I can see three: virtual channels (done),
> proper interpolation, and mapping data between head positions. The last two
> should be fairly straightforward if the generalization is good.
>
cc @yousrabk who will give the third application a shot :)


> My vote is to try to get all three done, and if it ends up being too
> ambitious, we can scale back later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1943/files#r28625950>.
>

Eric89GXL commented on mne/forward/_lead_dots.py:
```diff
@@ -253,9 +297,69 @@ def _do_self_dots_subset(intrad, rmags, rlens, cosmags, ws, volume, lut,
+            res = _fast_sphere_dot_r0(intrad, rmags1[ci1], rmags2[ci2],
+                                      rlens1[ci1], rlens2[ci2], cosmags1[ci1],
+                                      cosmags2[ci2], ws1[ci1], ws2[ci2],
+                                      volume, lut, n_fact, ch_type)
```
What is the speed like for this? I have a feeling we might be able to use my `np.bincount` trick to get rid of at least one of the loops.

mainakjas replied:
actually, it's pretty fast on my computer. I'm a bit confused why you needed to parallelize. Or maybe my computer is too powerful ...

agramfort replied:
it's slow if you interpolate on high resolution mesh (helmet or scalp) but
not here.

Eric89GXL replied:
Yeah I think the scalp surface was the tough one, even the BEM version to ~2000 vertices, the dense one a couple orders of magnitude more.

mainakjas replied:
makes sense

Eric89GXL commented on mne/forward/_field_interpolation.py:
```diff
@@ -87,6 +107,84 @@ def _compute_mapping_matrix(fmd, info):
+    ch_names = [c['ch_name'] for c in info_from['chs']]
+    fmd = dict(kind=ch_type, ch_names=ch_names,
+               origin=my_origin, noise=noise, self_dots=self_dots,
+               surface_dots=cross_dots, int_rad=int_rad, miss=miss)
```
How much of this code is shared with the make_field_map code? It would be good to stay as DRY as possible and unify where we can.

Eric89GXL replied:
In my head `make_field_map` is just a particular case where we want to map from our `N` sensors onto virtual mags at `M` surface positions.

Eric89GXL commented:
I think of Matti's C code as a treasure trove embedded in a sea of C boilerplate. Once you find what you're looking for, the code is indeed quite clear.

I can't decide if I'm more excited about the fact that it now seems to be working, or that there is now someone else who we can bother to help translate Matti's code when necessary :)

Eric89GXL commented on examples/plot_virtual_evoked.py:
```diff
@@ -0,0 +1,66 @@
+evoked = epochs.average()
+
+# go from grad to mag
+from mne import layouts
```
This will make Travis angry, move import to top

mainakjas replied:
oh we don't actually need it any more. I'll remove this

Eric89GXL commented on mne/forward/_lead_dots.py:
```diff
@@ -253,9 +297,69 @@ def _do_self_dots_subset(intrad, rmags, rlens, cosmags, ws, volume, lut,
 
 
+def _do_cross_dots(intrad, volume, coils1, coils2, r0, ch_type,
+                   lut, n_fact):
```
Please add the file where you found this -- I might have forgetten to do that with some of these so I recognize this is a hypocritical critique :) If you feel like adding the ones that I forgot, feel free, but you don't have to. It's just nice for the next person to have the trail of breadcrumbs to follow.

Eric89GXL replied:
...don't forget to leave the trail back to mne-C for the next treasure hunter.

Eric89GXL replied:
and by that, I mean a filename preferably. I see you have `mne_map_data` below, I assume that doesn't mean `mne_map_data.c` but the `mne_map_data` source code. I assume this actually lives in the mne-libs/forward somewhere...? If it is in fact `mne_map_data.c`, just add the file extension below.

agramfort commented:
now that the code works we need to agree on API for the different use cases. @mainakjas can you make a proposition with snippets? thx

mainakjas commented:
For virtual evoked, we can do:

`evoked = evoked.compute_virtual_evoked(to_type='mag')`

which will map all the existing MEG channels to `mag` locations in `evoked.info` and return an evoked with 102 channels. 'to_type' can be either 'mag' or 'grad'.

For bad channel interpolation, we can simply do:

`evoked = evoked.interpolate_bads_meg()`

For head position, I am not sure. But maybe the user can provide a `trans` file/matrix between the two head positions?

`evoked = evoked.change_head_position(trans)`

Wdyt?

agramfort commented:
> For virtual evoked, we can do:
>
> evoked = evoked.compute_virtual_evoked(to_type='mag')

I would say.

evoked = evoked.as_type('mag')

a method of evoked that contains evoked is unnecessary.

> For bad channel interpolation, we can simply do:
>
> evoked = evoked.interpolate_bads_meg()

+1

and we could have an

evoked.interpolate_bads()

to repair all channels.

> For head position, I am not sure. But maybe the user can provide a trans file/matrix between the two head positions?
>
> evoked = evoked.change_head_position(trans)

I would use a function here that takes a list of evoked and puts them
all in the same head position.

my 2c

mainakjas commented:
On Sat, Apr 18, 2015 at 1:09 PM, Alexandre Gramfort <
notifications@github.com> wrote:

> > For virtual evoked, we can do:
> >
> > evoked = evoked.compute_virtual_evoked(to_type='mag')
>
> I would say.
>
> evoked = evoked.as_type('mag')
>
> a method of evoked that contains evoked is unnecessary.
>

+1 I like this.


> > For bad channel interpolation, we can simply do:
> >
> > evoked = evoked.interpolate_bads_meg()
>
> +1
>
> and we could have an
>
> evoked.interpolate_bads()
>
> to repair all channels.
>

I knew that was coming :)


> > For head position, I am not sure. But maybe the user can provide a trans
> file/matrix between the two head positions?
> >
> > evoked = evoked.change_head_position(trans)
>
> I would use a function here that takes a list of evoked and puts them
> all in the same head position.
>

makes sense. So, it would put them all in the head position of the first
evoked of the list?


> my 2c
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1943#issuecomment-94156012>.
>

agramfort commented:
>
> makes sense. So, it would put them all in the head position of the first
> evoked of the list?
>

+1

mainakjas commented:
are we happy with the example now? I'll add tests if the example looks good. I'll remove the `add_title` function in the example once @cmoutard's colorbar fix is merged.

Eric89GXL commented:
Eventually we probably do want to be able to specify the head position instead of just taking the first one, since for example we translate all head positions to the median (in quaternion space) head position when doing translations. If `trans=None` then use first head position. Sound reasonable?

agramfort commented on examples/plot_virtual_evoked.py:
```diff
@@ -0,0 +1,47 @@
+add_title('mag (interpolated from mag + grad)')
+
+# go from grad + mag to grad
+virt_evoked = evoked.as_type('grad', copy=True)
```
I would not allow a copy param here and return a new instance all the time

agramfort commented on mne/evoked.py:
```diff
@@ -707,6 +707,29 @@ def as_data_frame(self, picks=None, scale_time=1e3, scalings=None,
+        to_type : str
+            The destination channel type. It can be 'mag' or 'grad'.
+        copy : bool
+            If False evoked is modified in place.
```
remove the copy param here.

my 2c

agramfort commented on mne/forward/_field_interpolation.py:
```diff
@@ -87,6 +105,84 @@ def _compute_mapping_matrix(fmd, info):
     return mapping_mat
 
 
+def _as_type_evoked(evoked, to_type='grad', mode='fast', copy=True, n_jobs=1):
```
I would name it _as_meg_type_evoked to make clear that it's for meg channels only

agramfort commented:
can you see what happens when you apply this to CTF data?

agramfort commented on mne/forward/_field_interpolation.py:
```diff
@@ -87,6 +105,84 @@ def _compute_mapping_matrix(fmd, info):
+
+    # keep only the destination channel types
+    evoked.pick_types(meg=to_type, eeg=False, ref_meg=False)
+    evoked.data = data
```
do we really want to keep the coil type unchanged here?

Eric89GXL replied:
I think we decided to keep the coil type the same (because they are supposed to act like actual coils), but change the channel names to make it clear it's interpolated

agramfort replied:
>
> I think we decided to keep the coil type the same (because they are
> supposed to act like actual coils), but change the channel names to make it
> clear it's interpolated
>
ok but I fear this will break layout auto detection :(

mainakjas replied:
if we change the coil type, won't it break picks?

On Sat, Apr 18, 2015 at 4:02 PM, Alexandre Gramfort <
notifications@github.com> wrote:

> In mne/forward/_field_interpolation.py
> <https://github.com/mne-tools/mne-python/pull/1943#discussion_r28644218>:
>
> > +               origin=my_origin, noise=noise, self_dots=self_dots,
> > +               surface_dots=cross_dots, int_rad=int_rad, miss=miss)
> > +    logger.info('Field mapping data ready')
> > +
> > +    #
> > +    # Step 3. Compute the mapping matrix
> > +    #
> > +    fmd['data'] = _compute_mapping_matrix(fmd, info_from)
> > +
> > +    # compute evoked data by multiplying by the 'gain matrix' from
> > +    # original sensors to virtual sensors
> > +    data = np.dot(fmd['data'], evoked.data[pick_from])
> > +
> > +    # keep only the destination channel types
> > +    evoked.pick_types(meg=to_type, eeg=False, ref_meg=False)
> > +    evoked.data = data
>
>  I think we decided to keep the coil type the same (because they are
> supposed to act like actual coils), but change the channel names to make it
> clear it's interpolated
> ok but I fear this will break layout auto detection :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1943/files#r28644218>.
>

mainakjas replied:
yeah, I agree. It became a pain with topomaps if we changed the channel names. I just keep the info of the `to_type` and replace the data.

agramfort replied:
I am just concerned if people solve the inverse problem based on
interpolated data. I don't know what to expect in terms of change...

Eric89GXL replied:
Can we store somewhere else that data were interpolated so that we can tell when someone tries to do an inverse?

I'm okay with putting a stern warning in the docstring, perhaps right after the one-line description.

agramfort commented on mne/forward/_lead_dots.py:
```diff
@@ -220,7 +281,30 @@ def _fast_sphere_dot_r0(r, rr1, rr2, lr1, lr2, cosmags1, cosmags2,
 
 def _do_self_dots(intrad, volume, coils, r0, ch_type, lut, n_fact, n_jobs):
-    """Perform the lead field dot product integrations"""
+    """Perform the lead field dot product integrations
```
missing empty line

agramfort commented on mne/forward/_lead_dots.py:
```diff
@@ -253,9 +338,89 @@ def _do_self_dots_subset(intrad, rmags, rlens, cosmags, ws, volume, lut,
+    sel : array
+        Indices of the surface vertices to select.
+    r0 : array, shape (3 x 1)
+        origin of the sphere.
```
origin -> Origin

Eric89GXL replied:
also `list of dict` not `list of dist` above

agramfort replied:
what if we append _virtual to channel names and remove it in
mne.utils._clean_names ? would plot topo work?

Eric89GXL replied:
+1 yeah that should work

mainakjas replied:
On Sat, Apr 18, 2015 at 8:08 PM, Alexandre Gramfort <
notifications@github.com> wrote:

> In mne/forward/_lead_dots.py
> <https://github.com/mne-tools/mne-python/pull/1943#discussion_r28646129>:
>
> > +
> > +    Parameters
> > +    ----------
> > +    intrad : float
> > +        The integration radius. It is used to calculate beta as:
> > +        beta = (intrad * intrad) / (r1 * r2)
> > +    volume : bool
> > +        If True, compute a volume integral.
> > +    coils : list of dist
> > +        The coils.
> > +    surf : dict
> > +        The surface on which the field is interpolated.
> > +    sel : array
> > +        Indices of the surface vertices to select.
> > +    r0 : array, shape (3 x 1)
> > +        origin of the sphere.
>
> what if we append _virtual to channel names and remove it in
> mne.utils._clean_names ? would plot topo work?
>

ok, this is done now. I am on to the tests now

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1943/files#r28646129>.
>

agramfort commented:
please add tests now. thanks @mainakjas !

mainakjas commented:
@agramfort topomap doesn't seem to be working on spm face CTF data. It isn't able to detect `grad` channels (even though picks finds `grad` channels). I think this should be dealt in a separate PR

mainakjas commented:
On Sat, Apr 18, 2015 at 3:46 PM, Eric Larson <notifications@github.com>
wrote:

> Eventually we probably do want to be able to specify the head position
> instead of just taking the first one, since for example we translate all
> head positions to the median (in quaternion space) head position when doing
> translations. If trans=None then use first head position. Sound
> reasonable?
>
yes, that sounds good.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1943#issuecomment-94165838>.
>

mainakjas commented:
When building the docs, I get this error: 

```py
Exception occurred:
  File "/home/mainak/anaconda/lib/python2.7/site-packages/pyface/ui/qt4/code_editor/pygments_highlighter.py", line 39, in get_tokens_unprocessed
    for item in action(self, m):
TypeError: 'NoneType' object is not callable
```

anyone else had this issue?

Eric89GXL commented:
I have been getting some stupid error related to `pygments` lately, too, might be that one :(

mainakjas commented:
I added tests and updated whats new. Only thing is that I am unable to build the docs -- travis seemed to have failed due to some doc issue. Otherwise, I am good to go

agramfort commented:
i tried

```
evoked.copy().pick_types(meg='grad').as_type('mag').plot_topomap(ch_type='mag')
```

and I get a cryptic error ""ValueError: No channels match the selection.". 
can you make it more explicit?

agramfort commented on examples/plot_virtual_evoked.py:
```diff
@@ -0,0 +1,35 @@
+
+In this example, MEG data are remapped from one
+channel type to another. This is useful for combining statistics
+for magnetometer and gradiometers.
```
This is useful to:

- visualize combined magnetometers and gradiometers as magnetometers or gradiometers
- run statistics from both magnetometers and gradiometers while working with a single type of channels

agramfort commented:
LGTM 

works great !

Eric89GXL commented on mne/forward/_field_interpolation.py:
```diff
@@ -87,6 +105,90 @@ def _compute_mapping_matrix(fmd, info):
     return mapping_mat
 
 
+def _as_meg_type_evoked(evoked, to_type='grad', mode='fast', n_jobs=1):
```
I think eventually we'll want to separate the evoked-ness of this from the remapping portion. Basically all you really need to do the remapping is an info struct and data. And even the data could be omitted and this function could return the remapping matrix to `dot` with potentially. It would make it so that we could more easily map data of any type (raw, epochs, evoked) if we want. But that can be saved for a later PR.

Eric89GXL commented on mne/forward/_lead_dots.py:
```diff
@@ -253,9 +341,96 @@ def _do_self_dots_subset(intrad, rmags, rlens, cosmags, ws, volume, lut,
+    cosmags2 = [coil['cosmag'] for coil in coils2]
+
+    products = np.zeros((len(rmags1), len(rmags2)))
+    for ci1, coil1 in enumerate(coils1):
```
WDYT about just doing `for rmag1, cosmag1, w1 in zip(rmags1, cosmags1, ws1):` here? You don't actually even use `coil1` at all, so it should be a little bit cleaner I think.

Eric89GXL replied:
ahh I guess you'd need to do `enumerate(zip)` which gets a bit long. This could at least be `for ci1 in range(len(coils1))`, then, since `coil1` isn't used.

Eric89GXL commented on mne/forward/tests/test_field_interpolation.py:
```diff
@@ -154,4 +159,45 @@ def test_make_field_map_meg():
+    cross_dots1 = _do_cross_dots(**args1)
+    cross_dots2 = _do_cross_dots(**args2)
+
+    assert_array_almost_equal(cross_dots1, cross_dots2.T)
```
I think there is another test you can do, which is a more transparent and practical one -- basically the tests you were doing visually. Use mags to interpolate to grads, compare to real grads. Use grads to interpolate to mags, compare to real mags. `np.allclose` will almost certainly not work since the accuracy isn't that high -- I'd see if a `np.corrcoef` over the `.ravel()`'ed result was e.g. 0.99 or higher.

Eric89GXL commented on mne/utils.py:
```diff
@@ -1545,6 +1545,8 @@ def _clean_names(names, remove_whitespace=False, before_dash=True):
             name = name.replace(' ', '')
         if '-' in name and before_dash:
             name = name.split('-')[0]
+        if '_virtual' in name:
```
do we always append it to the end? If so, it might be safer to do `name.endswith()` and `name = name[:-8]`

Eric89GXL commented:
My comments are mostly nitpicks on what seems like great code. Otherwise LGTM!

agramfort commented:
@mainakjas address @Eric89GXL's comments than +1 for merge !

next PR will be interpolation of bad meg channels !

great job @mainakjas !

mainakjas commented:
all comments addressed @agramfort . I removed `n_jobs` for now because it doesn't help much with few channels + it's not implemented for `_do_cross_dots` -- so, it's only half the job (literally :)

dengemann commented on mne/evoked.py:
```diff
@@ -711,6 +711,31 @@ def as_data_frame(self, picks=None, scale_time=1e3, scalings=None,
+        Parameters
+        ----------
+        to_type : str
+            The destination channel type. It can be 'mag' or 'grad'.
```
as_type, to_type ... maybe ch_type as elsewhere?

dengemann commented on mne/forward/_field_interpolation.py:
```diff
@@ -87,6 +105,95 @@ def _compute_mapping_matrix(fmd, info):
+
+    # change channel names to emphasize they contain interpolated data
+    for ch in evoked.info['chs']:
+        ch['ch_name'] = ch['ch_name'] + '_virtual'
```
why not inplace? +=

dengemann commented on mne/forward/_lead_dots.py:
```diff
@@ -139,7 +145,26 @@ def _comp_sum_eeg(beta, ctheta, lut_fun, n_fact):
+
+    Parameters
+    ----------
+    beta : array, shape (n_points * n_points, 1)
```
type is missing

dengemann commented on mne/forward/_lead_dots.py:
```diff
@@ -139,7 +145,26 @@ def _comp_sum_eeg(beta, ctheta, lut_fun, n_fact):
+    ----------
+    beta : array, shape (n_points * n_points, 1)
+        Coefficients of the integration.
+    ctheta : array, shape (n_points * n_points, 1)
```
same here

dengemann commented on mne/forward/_lead_dots.py:
```diff
@@ -139,7 +145,26 @@ def _comp_sum_eeg(beta, ctheta, lut_fun, n_fact):
+        Coefficients of the integration.
+    ctheta : array, shape (n_points * n_points, 1)
+        Cosine of the angle between the sensor integration points.
+    lut_fun : object
```
? really? nothing more specific?

mainakjas replied:
that's the best I could figure out :) maybe @Eric89GXL can give a hand with the docs here.

Eric89GXL replied:
IIRC it's a callable function for the lookup table (LUT)

dengemann commented on mne/forward/_lead_dots.py:
```diff
@@ -139,7 +145,26 @@ def _comp_sum_eeg(beta, ctheta, lut_fun, n_fact):
+
+    Returns
+    -------
+    sums : array, shape (4, n_points * n_points)
```
type

Eric89GXL commented on mne/forward/tests/test_field_interpolation.py:
```diff
@@ -154,4 +159,53 @@ def test_make_field_map_meg():
+    assert_array_almost_equal(cross_dots1, cross_dots2.T)
+
+    # correlation test
+    evoked = evoked.pick_channels(ch_names=ch_names[:10:]).copy()
```
I'm actually surprised/impressed that it works with so few channels, I figured you'd need more than that to get a good re-interpolation. But I guess it works beacuse they are all already in the right position.

mainakjas replied:
That makes sense, but that's why I need to understand what's actually going on for the interpolation :)

dengemann commented on mne/forward/_lead_dots.py:
```diff
@@ -162,7 +187,43 @@ def _comp_sums_meg(beta, ctheta, lut_fun, n_fact, volume_integral):
+    lr1 : array, shape (n_points x 1)
+        Magnitude of position vector of integration points in first sensor.
+    lr2 : array, shape (n_points x 1)
+        Magnitude of position vector of integration points in second sensor.
```
types are missing for array docs

Eric89GXL replied:
Is that another standard we're adding? I haven't seen us do that most places. In any case these are private so I think it's okay.

mainakjas replied:
yeah, I copied this style from `epochs.py` but I'm happy to change if we all agree

dengemann replied:
We're inconsistent here. Lately we were more strict about it, adopting sklearn conventions. If you're lazy keep it for a big PR once some one is bored.

dengemann commented on mne/forward/tests/test_field_interpolation.py:
```diff
@@ -154,4 +159,53 @@ def test_make_field_map_meg():
     assert_raises(ValueError, make_field_map, evoked, ch_type='foobar')
 
 
+def _setup_args(info):
```
"""helper to ..."""

dengemann commented:
LGTM but I'd actually prefer something like ch_type as parameter name. reading as_type and then to_type somehow distracts me. Also this would be a name that we already have in the context of selecting channels for plotting. What do others think?

mainakjas commented:
I agree. I can change it to `ch_type`

Eric89GXL commented:
I'm fine with `ch_type`

dengemann commented:
cool +3 on that already, @agramfort +0 or +1 and good?

mainakjas commented:
@dengemann any other comments? I'll address them if not

dengemann commented:
@mainakjas other wise it looks good to me!

dengemann commented:
Really nice work!

mainakjas commented:
Other than the comment about types for array docs (which I feel lazy to fix), others are taken care of.

Eric89GXL commented:
Merged, thanks @mainakjas!

agramfort commented:
great job @mainakjas !
